### ISSUE TITLE
[RFC]: memory allocation and fragmentation

### ISSUE THREAD

**Issue by rianquinn** [Bareflank/hypervisor#425]
### Description

The current memory manager has a couple of issues that need to be addressed.

- Fragmentation is terrible with the current implementation. Any long term use of malloc / free will eventually result in failures. A SLAB style allocator is needed to address this problem. 
- There is no ability to expand the current memory allocation. This will not scale when attempting to allocate large amounts of memory in the VMM, or attempting to launch guests with large memory needs. 
- The current memory manager stores virt->phys mappings in a set of maps, instead of looking up this information from a set of page tables. This results in wasted memory and added complexity. 
- The current memory manager is labeled xxx_x64, which means it's specific to x86_64, and it doesn't need to be. 

There are also some interesting constraints that Bareflank has to deal with. 

- The VMM cannot modify the page tables while bootstrapping (i.e. prior to the first VM exit). Therefore, it has no means to "map" memory while bootstrapping. 
- The VMM has no control over where it is placed in memory. 

Our goal, under these constraints, is to update the memory manager such that most fragmentation issues are resolved, while reducing memory overhead / complexity and providing a means to increase the amount of memory the hypervisor has access to.  

### Bootstrap Memory:

The first allocation to occur happens by libcxx / the VMM while executing global constructors (for example, the VMCS has a global GDT that performs an allocation, same with the IDT and the TSS). We are only given one choice for providing this malloc'd memory, which is to use a "placement new" scheme, which is exactly what we do today. We take a contiguous buffer in memory and begin to divide it up. 

To perform this allocation we will use a simple best-fit allocator (fragmentation will be addressed so please bear with me). A single contiguous memory block will be dedicated in the memory manager's ".bss" section as follows:

```
--------------------
| free_list        |
|                  |
--------------------
--------------------
| used_list        |
|                  |
--------------------
--------------------
| start            |
|                  |
|                  |
|                  |
|                  |
|                  |
|                  |
|                  |
|                  |
|                  |
|                  |
|                  |
|                  |
--------------------
```

The following struct will also be needed:

```
struct block_t
{
    void* addr;
    uintptr_t size;
    block_t *next;
    block_t *prev;
};
```

As shown above, there are three memory regions that need to be maintained using four different linked lists. Unlike most allocators, the block struct will **not** be stored above the memory that is being allocated. This is because the hypervisor frequently needs to allocate aligned memory. Specifically, page aligned memory. For this reason, the first page will be used for the free list. Once the free list has consumed half of this page, the free list will allocate from itself, another page. The pages used by the free list will be maintained in a second list called the free_list_list (the recursive name is intentional as it's abrasive). 

Notes:
- All allocations from the free_list must be a multiple of a page, and must also be page aligned. 
- How smaller allocations are handled, as well as fragmentation will be discussed later. 

When malloc() is run, the block in the free_list that best fits the requested size will be used or split up (if the block is larger than the requested size). The current free_list entry will be updated to reflect the new addr / size of the remaining memory. The consumed memory will be added to a third used_list. The used_list will have the same properties of the free_list. A page will be dedicated for the used_list, and once half of the used_list is consumed, a page will be allocated from itself and added to the used_list_list. 

Notes:
- When free is called, the used_list will be scanned for the address that is provided which should result in the size of the memory allocated. The block will then be moved from the used_list to the free_list. 
- When adding blocks to the free list, the memory will be perpended, not appended. This is because allocations tend to free / alloc the same size, and thus should result in a mild optimization (i.e. the quick fit optimization)
- If free is given an address in the middle of a previously allocated block, a warning will be toss up, but the block will still be freed. This should never happen, but might if a destructor is not marked virtual. 
- If free is given an address that is totally bogus, we will toss up a warning and ignore the request. 

Since both the free_list and the used_list have fixed sized blocks (the size of struct block, which intentionally is a cache line), fragmentation is not an issue. We do however need to worry about how to reused blocks as they are consumed / released. To address this, when memory from a block in the free_list is consumed to 0, the block will be left with an addr and size of 0. When a split occurs on a free_block, the allocator will look for a 0 sized block before creating a new one. That is, links in the linked list will never be removed, but instead be left in a state where they can be reused later. The same approach will be used for the used_list. When memory is freed, the allocator will look for a block in the free_list that is size 0 before creating a new one and will set the block in the used_list to 0. When memory is allocated, the allocator will look for blocks in the used_list with size 0 before creating a new block in the list. Using this approach fragmentation in the linked lists should be avoided. After all of this is working, we can go back and look at possibly reclaiming unused pages in free_list_list and used_list_list by consolidating the linked list if enough fragmentation has occurred such that a page can be reclaimed (a case that is not likely to be an issue for now, but easy to address in the future with this design).

### Expanding Memory

The current implementation has a fixed size block of memory for allocations. This is not scalable, and we have already run into issues with users running out of memory (or having to allocate huge amounts of memory to deal with fragmentation). In addition, the VMM cannot map memory during bootstrapping, but it can once the VMM is up and running via a VMcall. Finally, we cannot control where the VMM is placed in memory. Thus, memory looks like this when we start:

```
--------------------
|                  |
|                  |
|                  |
|                  |
|                  |
|------------------|
| VMM              |
|------------------|
|                  |
|                  |
|                  |
|                  |
|                  |
--------------------
```

To solve these problems, the best fit algorithm will provide a means to increase it's memory. To accomplish this, the algorithm will be given the top and bottom of the VMM by the driver (new vmcalls that will be added). Then, the driver can add pages to the hypervisor whenever it wants (we will even extend BFM to tell the bfdriver to add more memory to the hypervisor which should prove to be a useful added feature, but also a great way to test this code). 

Virtual memory has the same problem with respect to fragmentation and allocation when dealing with mapping. To solve this, a virtual memory manager will be added that uses the same algorithm as the bootstrap allocator. This way virtual memory can be allocated / freed for mapping. When memory needs to be added to the bootstrap allocator, virtual memory is allocated, and mapped to a set of pages provided by the host OS for expansion. This memory is then added to the free_list in the bootstrap allocator for future allocations. Reclaiming memory from the VMM is as simple as going through the free_list, identifying pages that are not being used, and removing them from the free_list, providing the host OS with the physical address of the pages so that they can be reclaimed by the host OS. 

### Fragmentation

Using a best-fit algorithm will result in fragmentation that is just as bad as what we have today. Additionally, since each allocation must be a multiple of a page, a lot of memory will be wasted with the above approach. To solve fragmentation, a memory pool algorithm that is either type aware, or uses fixed sized blocks must be used. What this means is that each time you malloc, if you know that your always going to allocate 10 bytes, you allocate 10 bytes, but when you free this memory you place it in a queue to be used the next time you alloc again. There are a lot of advantages to this approach:

- If you do this approach for all allocations, you completely remove fragmentation. 
- Allocations / freeing memory is O(1) which is ideal (first-fit, more on this)
- Cache alignment / management is much easier with type aware allocators. 

The Linux kernel uses this approach named the SLAB / SLUB allocator. C++ however makes this process a lot easier for us. Each class can have it's new / delete operators overloaded providing a means to implement a type aware allocator. An example of this can be found [here](https://github.com/cacay/MemoryPool). A memory pool allocator that adheres to the [allocator concept](http://en.cppreference.com/w/cpp/concept/Allocator) provides a simple means to overload the new / delete operators for each type, or provide custom allocators to std::xxx data structures (e.g. for basic types if needed). 

To start, lets review the vCPU type. 
A type like a vCPU has a constant size. The vCPU's new and delete operator will be overloaded using the following:

```cpp

memory_allocator<vcpu_intel_x64> g_ma_vcpu_intel_x64;

class vcpu_intel_x64 : public vcpu
{
    ....

    static void *operator new(size_t bytes)
    {
        return g_ma_vcpu_intel_x64.allocate();
    }

    static void *operator delete(void* ptr, std::size_t)
    {
        return g_ma_vcpu_intel_x64.deallocate(ptr);
    }
}
```

When a memory allocator is created, it will have an empty queue. Once the first allocation occurs, malloc() will be called getting a page. The allocator will maintain a free_list and a used_list, as well as it's own free_list_list and used_list_list of pages just like the bootstrap allocator. The difference is, the memory allocator will only be allocating fixed sized blocks using a first-fit algorithm which means that fragmentation will not occur. Cache alignment is optional. Since the design of the allocators are so similar, we will attempt to share as much code as possible. 

This approach can be used for all types defined in Bareflank and it's extension repos, greatly reducing fragmentation. The goal here is that malloc() uses a best-fit memory allocator, only providing memory in multiples of a page. First-fit memory allocators are created for each type, and divide up pages into fixed size chucks specific to the type they are serving. 

There is one issue with this approach, and that is with variable length allocations from things like JSON and std::string. To solve this problem, fixed sized allocators will be used where possible (for example, JSON vmcalls will used a fixed sized buffer of memory), and allocations that cannot be solved this way will give out memory in multiples of a page, wasting memory, but preventing fragmentation. 

Finally, debugging logic will be added to the allocators, providing a means to see what size allocations are being requested. This will help to identify when an allocation occurs that should have a custom new / delete operator, or an additional memory allocator added to the bootstrap allocator for commonly used sizes that are not exactly a page (for example, STACK_SIZE and JSON fixed size buffer that are 16K-64K in size). 

To test out fragmentation, we will be testing these situations using Jekins to prove that allocations are working for a LONG time. The tests that we plan to run are (please provide feedback if you would like to see other tests):
- mapping large amounts of guest memory over and over
- starting / stopping the hypervisor over and over
- register based VMcalls over and over
- json based VMcalls over and over

### sizeof(T)

The goal of this class is to handle any type. There are two different ways the allocator will end up with memory. If the allocator is the bootstrap allocator, it gets added large, contiguous blocks of memory by the bfdriver (i.e. contiguous blocks of virtual memory are mapped to a set of pages and then added to the free list of the bootstrap allocator). All other allocators get their memory from the bootstrap allocator via ::new() / ::delete(). 

If sizeof(T) == 8 bytes == sizeof(uint64_t), each allocation consumes 40 bytes (8 + sizeof(block)). 512 can fit in a single page evenly, which means that there is no waste. Overhead == 400%. 

If sizeof(T) == 64 bytes == cache line, each allocation consumes 96 bytes (64 + sizeof(block)). 64 can fit in a single page evenly, which means that there is no waste. Overhead == 50%. As can be seen, as the size of T increases, the overhead decreases. This relationship doesn't last forever. 

If sizeof(T) == 2048 bytes, each allocation consumes 2080 bytes (2048 + sizeof(block)). 2 can fit in a single page evenly, which means that there is no waste. Overhead == 1.5%. Now, lets see what happens when we add one more byte.

If sizeof(T) == 2049 bytes, each allocation consumes 2080 bytes (2048 + sizeof(block)). 1 can fit in a single page, with 2047 wasted. Memory is wasted because we only allocated a single page, and we cannot assume that allocating another page will result in virtual continuity. Because of this, any allocation above 2048 consumes a full page. Interestingly, the overhead is only 101.5%. Depending on how many of these are allocated, even though the overhead is low, the amount of memory that is wasted could be really high. 

What happens if sizeof(T) is larger than a page? The only solution to this problem is to allocate more than one page (allocation from the bootstrap allocator have to be in multiples of a page as that's all it's going to do). If we can allocate more than a single page, we should be able to reduce the overall overhead. Lets look at this. 

If sizeof(T) == 2049 bytes, each allocation consumes 2080 bytes (2048 + sizeof(block)). 3 can fit in two contiguous pages, with 2045 wasted. Overhead == 33%. As we increase the number of pages we allocate at a time, we reduce the total amount of memory that is wasted. 

Based on these results, the template arguments to the class will consist of the following:
- T: the type that is being managed, which will be used to get sizeof(T)
- BS: the block size. This will default to 0x1000. We are only adding this because other people might want to use our allocator, and they should not be constrained to a single page. 
- MIN_MULT: The minimum number of elements per allocation. This will default to 4. Specifically, as was shown above, making sure that at least 4 T can fit in a single allocation ensures that the overhead is low (on the same order as a cache line which is our average allocation size). 
- MIN_SIZE: This will define the minimum allocation regardless of the sizeof(T). This will default to sizeof(T), but can be changed to control cache alignment. For example, suppose sizeof(T) == 12. Accesses to this memory will thrash the cache more than if 16 bytes were used instead. 

It should be noted that a special case will be provided that covers the sizeof(T) == 4096. In this case, there is no waste, and thus a single allocation is optimal as it prevents fragmentation of the bootstrap allocator (remember, we care more about fragmentation and less about waste or speed). 

### Virt -> Phys

Finally, we need to deal with virt->phys mappings. This is a chicken and egg problem. We want to get our mappings from the same set of pages tables that we are about to use / are using. The problem is, how do you construct a set of page tables without the virt->phys mappings (and around and around we go). 

To solve this we will need to create a temporary data structure that stores the virt->phys mappings for the memory that is used by the bootstrap allocator. This temporary structure will be a std::map containing the mappings as needed. The bfdriver will provide the mappings for the bootstrap allocator memory first. Once this information is available, the bfdriver will tell the VMM to construct it's CR3. When virt->phys is called, the memmory manager will use the temporary virt->phys map to handle the conversions. Once the CR3 is constructed, these temporary maps will be cleared, and future requests will come from CR3. Since malloc() only provided multiples of a page, the memory used by the temporary map will be released and consumed in the future without fragmentation. 

@tklengyel @ktemkin @brendank310 @josephsharkey @carter-yagemann @mor619dx

**Comment by rianquinn** [comment]
I updated the block struct to be the following:
```cpp
struct block_t
{
    void* addr;
    uintptr_t size;
    block_t *next;
    block_t *prev;
};
```

Using uintptr_t for the pointer types is not Core Guidelines compliant. Also, we don't need a flags field because we maintain a used and a free list, but making this a doubly linked list greatly simplifies things. 

Also, the specs for the allocator type can be found here:
http://en.cppreference.com/w/cpp/memory/allocator_traits
http://en.cppreference.com/w/cpp/concept/Allocator
http://en.cppreference.com/w/cpp/memory/allocator

All three are needed to fully understand how this works, and it should be noted that the reference implementation that I linked has a lot of incompatibilities with these specs so our implementation will likely not be the same.
